lib/auth/iam_join.go

```diff
@@ -0,0 +1,245 @@
+		return trace.AccessDenied("sts identity request does not include challenge header or it does not match")
+	}
+
+	authHeader := req.Header.Get(aws.AuthorizationHeader)
+
+	sigV4, err := aws.ParseSigV4(authHeader)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if !utils.SliceContainsStr(sigV4.SignedHeaders, normalizedChallengeHeaderKey) {
+		return trace.AccessDenied("sts identity request auth header %q does not include "+
+			normalizedChallengeHeaderKey+" as a signed header", authHeader)
+	}
+
+	body, err := aws.GetAndReplaceReqBody(req)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if !bytes.Equal([]byte(expectedSTSIdentityRequestBody), body) {
+		return trace.BadParameter("sts request body %q does not equal expected %q", string(body), expectedSTSIdentityRequestBody)
+	}
+
+	return nil
+}
+
+func parseSTSRequest(req []byte) (*http.Request, error) {
+	httpReq, err := http.ReadRequest(bufio.NewReader(bytes.NewReader(req)))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Unset RequestURI and set req.URL instead (necessary quirk of sending a
+	// request parsed by http.ReadRequest). Also, force https here.
+	if httpReq.RequestURI != "/" {
+		return nil, trace.AccessDenied("unexpected sts identity request URI: %q", httpReq.RequestURI)
+	}
+	httpReq.RequestURI = ""
+	httpReq.URL = &url.URL{
+		Scheme: "https",
+		Host:   stsHost,
+	}
+	return httpReq, nil
+}
+
+type stsIdentityResponse struct {
+	GetCallerIdentityResponse struct {
+		GetCallerIdentityResult awsIdentity
+	}
+}
+
+type awsIdentity struct {
+	Account string
+	Arn     string
+}
+
+type stsClient interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+func executeSTSIdentityRequest(ctx context.Context, client stsClient, req *http.Request) (identity awsIdentity, err error) {
+	req = req.WithContext(ctx)
+	resp, err := client.Do(req)
+	if err != nil {
+		return identity, trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return identity, trace.Wrap(err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return identity, trace.AccessDenied("aws sts api returned status: %q body: %q",
+			resp.Status, body)
+	}
+
+	var identityResponse stsIdentityResponse
+	if err := json.Unmarshal(body, &identityResponse); err != nil {
+		return identity, trace.Wrap(err)
+	}
+	return identityResponse.GetCallerIdentityResponse.GetCallerIdentityResult, nil
+}
+
+// arnMatches returns true if arn matches the pattern. pattern should be an AWS
+// ARN which may include "*" to match any combination of zero or more characters
+// and "?" to match any single character, see https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_resource.html
+func arnMatches(pattern, arn string) (bool, error) {
+	// asterisk should match zero or
+	pattern = strings.ReplaceAll(pattern, "*", ".*")
+	pattern = strings.ReplaceAll(pattern, "?", ".")
+	pattern = "^" + pattern + "$"
+	matched, err := regexp.MatchString(pattern, arn)
```

Comment:
Any regex-like expression in `pattern` are leaking to the final comparison. Let's try something like this:

```suggestion
	pattern = regexp.QuoteMeta(pattern)
	pattern = strings.ReplaceAll(pattern, "\\*", ".*")
	pattern = strings.ReplaceAll(pattern, "\\?", ".")
	pattern = "^" + pattern + "$"
	matched, err := regexp.MatchString(pattern, arn)
```

A proper glob-matching function would be better, but none occur to me right now.

Reply:
Yeah true, I wasn't too worried about it because this is set by the admin (and could be sort of a shadow-feature allowing them to put more complex regexes) but probably better to block it. I followed your suggestion